### DISCUSSION TITLE
Visual testing

### DISCUSSION THREAD

**Discussion by Newbie012** [vitest-dev/vitest#690] [Ideas]
A third-party jest matcher called [`toMatchImageSnapshot`](https://github.com/americanexpress/jest-image-snapshot) performs image comparison. It's primarily useful when you do visual tests. It would be pretty cool if vitest could support this kind of testing as one of its built-in features.

The lack of visual testing prevents me from migrating from jest to vitest. I believe many other developers who perform visual testing will be interested in this feature.

**Comment by sheremet-va** [comment]
This library should work with Vitest, no?

**Reply by Newbie012** [reply]
I haven't really tried. After inspecting the source code, it looks like it really should. I'll keep you posted  :)

**Comment by hcampos-professional** [comment]
Agreed 👍. It would be great to support something similar to what Playwright has built in: https://playwright.dev/docs/test-snapshots

With Vitest we can take the screenshots but don't have a built in mechanism for comparing and asserting them. And even if we implement our own comparison (e.g. with [pixelmatch](https://github.com/mapbox/pixelmatch) or [phash](https://www.phash.org/)) there is no way (that I could find, at least) to display results in the Vitest UI. At the moment we're forced to build a custom reporter that produces a separate report that developers need to check separately.

It would be awesome if we could pass some metadata along with errors to let Vitest display `Before`, `After` and `Diff`, like Playwright does:

![image](https://github.com/user-attachments/assets/1a310b5d-9dc7-409c-8a63-b51e07a7b059)

Vitest already has something similar, in the form of `meta.failScreenshotPath`:

https://github.com/vitest-dev/vitest/blob/68e6c634c50e303d2adb968eae748d01317e2722/packages/ui/client/components/views/ViewReport.vue#L117-L123

Alternatively (or in addition), Vitest could add support for rendering HTML error messages. That would let apps render before/after/diff directly in the UI. Probably a way of opting out of escaping HTML here:

https://github.com/vitest-dev/vitest/blob/68e6c634c50e303d2adb968eae748d01317e2722/packages/ui/client/components/views/ViewReport.vue#L33-L61

**Comment by Shiwa** [comment]
I agree it would be pretty cool to do visual regression testing just using vitest built-ins.

I’m working on a project with lots of home made components and design that can be technically usable but visually broken (popovers, sticky toolbars, responsive layouts, etc.).
We already use chromatic for our design system and it’s awesome, and I’d really like to have the same kind of tooling in the app, without setting up extra libs.

**Comment by Shiwa** [comment]
FYI I implemented visual regression testing with vitest browser mode / playwright. It was painful and the current solution is not perfect.

Here are the pain points I encountered:
- the jest-image-snapshot does not work in browser mode, because it tries to use `Buffer`. I tried to use it in a server-side command but in this context `expect` does not work
- the `expect().toHaveScreenshot()` utility from playwright does not work because we don’t have access to the playwright version of `expect`
- so I used `page.screenshot()` and `toMatchSnapshot()` on the base64. This works, but the match must be pixel perfect.
- since the base64 snap is not very practical, I tried to use `readFile` and `toMatchFileSnapshot(ctx.task.name + '.png')`, but even with `encoding: 'binary'` the resulting snapshot is not a valid png file. I ended up doing a `readFile`/`writeFile` to save the png.

Here’s the current version of our visual regression testing utility, if that can help:

```
export async function renderAndScreenshot(
  element: JSX.Element,
  ctx: TaskContext,
) {
  renderInBrowser(element, {
    wrapper: ({ children }) => <BrowserTestContext>{children}</BrowserTestContext>, // BrowserTestContext imports CSS and stuff
  });

  const screenshotPath = `./__screenshots__/${ctx.task.name}.png`;
  const screenshot = await page.screenshot({ base64: true, path: screenshotPath });

  await expect(screenshot.base64).toMatchFileSnapshot(`./__snapshots__/${ctx.task.name}.snap`); // match base64

  // test ok, save png in __snapshots__ for reference
  const data = await server.commands.readFile(screenshot.path, { encoding: 'binary' });
  await server.commands.writeFile(`./__snapshots__/${ctx.task.name}.png`, data, { encoding: 'binary' });
  await server.commands.removeFile(screenshot.path);
}
```

**Comment by JoaoCnh** [comment]
Yeah I believe if we get something like this working within Vitest a lot of migration can happen.

We're currently already using Vitest but we would really love to have some regression testing completely integrated into vitest but we've not been successful with jest-image-snapshot or any tries with playwright custom commands.

**Comment by JoaoCnh** [comment]
I've tried writing my own custom matcher but a problem I can't seem to get around is having to use `fs` and `path` for file related stuff to handle screenshots and snapshots. I will always get hit with the `Module "path" has been externalized for browser compatibility. Cannot access "path.join" in client code.` error.

Is there a way to get around this?

**Reply by JoaoCnh** [reply]
I'm using Vitest browser testing btw

**Reply by Shiwa** [reply]
You should do file system actions through server commands, either using the provided `readFile`, etc. (like in my prototype above) or creating a custom one.
Custom server commands receive a context object with a `testPath` prop that helps to find your screenshots.

**Reply by JoaoCnh** [reply]
I was able to save the screenshot with success but pngjs then cannot read the contents from the `readFile` server command so I guess I'm hitting some sort of limitation of data, buffers and stuff that I'm not fully knowledgeable of. But maybe I can create a custom command to run this stuff on the server for me. I'll try that approach next

**Reply by JoaoCnh** [reply]
I was able to get it working. Now I need to fine-tune the logic of file organisation and all of that. Essentially the test page takes the screenshot and saves it -> passes the path to expect that was extended with a custom matcher called toMatchImageSnapshot -> that in turn calls a custom command that receives the test name and the screenshot path and does file reading/writing/deleting.

**Reply by svozza** [reply]
Thank you for the inspiration @JoaoCnh! For anyone interested in the code, here's how I got my image diffs working:

Here's the server command:

```js
import {promises as fs} from 'node:fs';
import path from 'node:path';
import pixelmatch from 'pixelmatch';
import {PNG} from 'pngjs';

/**
 * @typedef {Object} CompareOptions
 * @property {string} testName - Name of the test for organizing snapshots
 * @property {string} [baselineDir] - Directory for baseline images
 * @property {string} [diffDir] - Directory for diff images
 * @property {number} [threshold] - Matching threshold (0-1)
 * @property {number} [maxDiffPercentage] - Maximum allowed difference percentage
 * @property {boolean} [updateBaseline] - Whether to update baseline images
 */

/**
 * Compare a screenshot with its baseline
 * @param {Object} context - Context
 * @param {string} screenshotPath - Path to the screenshot
 * @param {CompareOptions} options - Comparison options
 * @returns {Promise<{matches: boolean, diffPercentage?: number, message: string}>}
 */
export const compareScreenshot = async (context, screenshotPath, options) => {
    const {testPath} = context;
    const testDir = testPath.replace(/\/[^/]*$/, '');
    const {
        testName,
        baselineDir = `${testDir}/__image_snapshots__`,
        diffDir = `${testDir}/__image_diffs__`,
        threshold = 0.1, // recommended default in pixelmatch docs
        maxDiffPercentage = 1.0,
        updateBaseline = process.env.UPDATE_SNAPSHOTS === 'true',
    } = options;

    const testBaselineDir = path.join(baselineDir, testName);
    const testDiffDir = path.join(diffDir, testName);

    const filename = path.basename(screenshotPath);
    const baselinePath = path.join(testBaselineDir, filename);
    const diffPath = path.join(testDiffDir, `diff-${filename}`);

    // Create directories if they don't exist
    await fs.mkdir(testBaselineDir, {recursive: true});
    await fs.mkdir(testDiffDir, {recursive: true});

    if (updateBaseline) {
        await fs.copyFile(screenshotPath, baselinePath);
        return {
            matches: true,
            message: `Updated baseline image: ${baselinePath}`,
        };
    }

    try {
        // Check if baseline exists
        await fs.access(baselinePath);

        const [img1Data, img2Data] = await Promise.all([
            fs.readFile(baselinePath),
            fs.readFile(screenshotPath),
        ]);

        const img1 = PNG.sync.read(img1Data);
        const img2 = PNG.sync.read(img2Data);

        if (img1.width !== img2.width || img1.height !== img2.height) {
            return {
                matches: false,
                message: `Image dimensions don't match: ${img1.width}x${img1.height} vs ${img2.width}x${img2.height}`,
            };
        }

        // Create empty diff image buffer (pixelmatch will mutate this buffer to store the diff)
        const {width, height} = img1;
        const diff = new PNG({width, height});

        // Compare images
        const numDiffPixels = pixelmatch(
            img1.data,
            img2.data,
            diff.data,
            width,
            height,
            {threshold}
        );

        // Save diff image
        await fs.writeFile(diffPath, PNG.sync.write(diff));

        const diffPercentage = (numDiffPixels / (width * height)) * 100;
        const matches = diffPercentage <= maxDiffPercentage;

        return {
            matches,
            diffPercentage,
            message: matches
                ? `Image matches baseline (diff: ${diffPercentage.toFixed(2)}%)`
                : `Image differs from baseline by ${diffPercentage.toFixed(2)}% (threshold: ${maxDiffPercentage}%). See diff: ${diffPath}`,
        };
    } catch (error) {
        if (error.code === 'ENOENT') {
            // Create baseline if it doesn't exist
            await fs.copyFile(screenshotPath, baselinePath);
            return {
                matches: true,
                message: `Created new baseline image: ${baselinePath}`,
            };
        }
        throw error;
    }
};
```

The custom matcher:

```js
import {expect} from 'vitest';
import {commands} from '@vitest/browser/context';

expect.extend({
    async toMatchImageSnapshot(received, options = {}) {
        // Get the current test name from Vitest's context
        const testPath = this.testPath.split('/');
        const testName = testPath.at(-1);

        // Handle different types of received values
        let screenshotPath;
        if (typeof received === 'string') {
            screenshotPath = received;
        } else if (received && typeof received === 'object' && received.path) {
            // Handle screenshot result object
            screenshotPath = received.path;
        } else {
            throw new Error(
                'Expected a screenshot path string or screenshot result object'
            );
        }

        const result = await commands.compareScreenshot(screenshotPath, {
            testName,
            ...options,
        });

        return {
            pass: result.matches,
            message: () => result.message,
        };
    },
});
```

The test part of my setup file:

```js
import {compareScreenshot} from './src/tests/vitest/commands/screenshot.js';

// ...
export default defineConfig({
   // ...
   test: {
            include: ['src/tests/browser/**/*.{test,spec}.js'],
            setupFiles: [
                './src/tests/vitest/setupFiles/matchers.js',
            ],
            name: 'browser',
            browser: {
                viewport: {
                    height: 2000,
                    width: 5000,
                },
                commands: {
                    compareScreenshot,
                },
                provider: 'playwright',
                enabled: true,
                headless: true,
                instances: [{browser: 'chromium'}],
            },
        },
}
```

And it's used in tests like so:

```js
        // i couldn't get the canvas element using the other locators, if anyone has 
        // had success i'd love to hear it
        const canvas = await screen.getByTestId('my-test-id'); 

        // wait for diagram animations on canvas to complete
        await sleep(2000);

        const screenshotPath = await canvas.screenshot({
            scale: 'device',
        });

        await expect(screenshotPath).toMatchImageSnapshot({
            maxDiffPercentage: 5.0,
        });
 ```

**Reply by nwidynski** [reply]
@svozza I just want to say thanks, this just made my day 🚀

**Reply by ImLunaHey** [reply]
@svozza are you getting errors show up when the expect fails? tried to implement this and im not. 🤔 

https://github.com/ImLunaHey/imlunahey.com/blob/main/src/elements/Button.test.tsx#L39-L48

edit: found the issue. i wasn't calling `await` on the expect. btw the repo i linked has this working in typescript with correct types. 🙏
